File: Java/src/main/java/org/onlineoptimisation/problem/benchmark/RastriginFunction.java
package org.onlineoptimisation.problem.benchmark;

import org.armadillojava.Arma;
import org.armadillojava.Col;
import org.armadillojava.Datum;

public class RastriginFunction extends BenchmarkProblem {

  public RastriginFunction(int numberOfDimensions) {
    super(numberOfDimensions);
  }

  @Override
  public double getObjectiveValueImplementation(Col parameter) {
    Col z = delta(10).times(asymmetric(0.2, oscillate(parameter)));
    return 10 * (_numberOfDimensions - Arma.accu(Arma.cos(z.times(2 * Datum.pi)))) + Arma.accu(Arma.square(z));
  }

}

File: src/optimisationProblem/benchmark/rastriginFunction.cpp
#include <optimisationProblem/benchmark/rastriginFunction.hpp>

#include <cmath>

namespace hop {
  RastriginFunction::RastriginFunction(const unsigned int &numberOfDimensions) : BenchmarkProblem(numberOfDimensions), _delta(getScaling(std::sqrt(10.0))) {

  }

  double RastriginFunction::getObjectiveValueImplementation(const arma::Col<double> &parameter) const {
    arma::Col<double> z = _delta % getAsymmetricTransformation(0.2, getOscillationTransformation(parameter - _translation));

    return 10.0 * (static_cast<double>(_numberOfDimensions) - arma::accu(arma::cos(2.0 * arma::datum::pi * z))) + std::pow(arma::norm(z), 2);
  }
}
